### PR TITLE
feat(docker-registry): use svc and rc

### PR DESCRIPTION
**Pull request by sgoings**

### CODE REVIEW

File: docker-registry/manifests/docker-registry-rc.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: registryrc
```
Comment (bacongobbler) [inline docker-registry/manifests/docker-registry-rc.yaml]: Let's keep the name as "registry" so its consistent with the package name.

**Comment by bacongobbler** [comment]
I recall there being a discussion that apps should provide the service rather than the provider. I'll see if I can find it :)

**Comment by bacongobbler** [comment]
Found it! https://github.com/deis/helm/blob/master/docs/modeling_services.md

**Comment by bacongobbler** [comment]
I completely understand the motivation but im still lost as to the "why" from k8s' standpoint. Kubernetes' docs does suggest using replication controllers instead of pods for prod... Even though these services aren't scalable. Pods have restart policies for when the app fails, so I'm still lost on why they're recommended since they seem to have all that we require to deploy a single, non-scalable app.

**Comment by slack** [comment]
I'm on the fence about consumers always being **required** to provide a service resource. In this case it feels entirely natural to include a service definition along with the registry components.

For apps that require a stable, configuration-free hostname; sure, bundle your own service definition. Otherwise I think it makes sense to include the service resource.

**Comment by bacongobbler** [comment]
It does leave a bad taste in my mouth to have (essentially) multiple service definitions that all point to the same provider. I tend to agree with @slack on this. I just wanted to bring it up :)

**Comment by sameersbn** [comment]
replication controllers allow you to add health checks which allows the controller the automatically restart a pod when the health checks fail. I think that is the main reason why you should use a replication controller even when the pod cannot be scaled.

**Comment by sameersbn** [comment]
okay it seems that pods definitions can also have heath checks.

**Comment by technosophos** [comment]
For registry, I think we should add a general service definition to it. Consumers can override, but since we're equally as likely to point Docker clients at it, we can't necessarily count on a consumer to provide the service definition.

As for the RC, @bacongobbler and @sameersbn are right. The documentation suggests using RCs instead of pods because...

> Unlike in the case where a user directly created pods, a replication controller replaces pods that are deleted or terminated for any reason, such as in the case of node failure or disruptive node maintenance, such as a kernel upgrade.

http://kubernetes.io/v1.1/docs/user-guide/replication-controller.html

But that's not really accurate, is it? I don't immediately see any reason why we'd select an RC for a Pod on a case where we know there can only ever be one.

**Comment by sgoings** [comment]
The main reason I chose to move it to an service+replication controller bundle is twofold:
- I get the assurance that there will always be a registry pod running in my cluster.
- the Kubernetes documentation points people to always use a replication controller even if you have a single-container-single-pod structure but want the thing to be long-lived.

@bacongobbler and others made me dive headfirst into the documentation to find out why I couldn't just rely on the `RestartPolicy = Always` reference within a pod.

In my trek through the documentation, I found that the a single container within a pod RestartPolicy is _close_ to being the same as one wrapped by a replication controller... except for one huge difference. The main advantage of a single-container-single-pod payload wrapped by a replication controller is that **with a replication controller the pod can be scheduled to another node on a node failure.**

See reference in the [Pod RestartPolicy documentation](http://kubernetes.io/v1.1/docs/user-guide/pod-states.html#restartpolicy):

> As discussed in the [pods document](http://kubernetes.io/v1.1/docs/user-guide/pods.html#durability-of-pods-or-lack-thereof), once bound to a node, a pod will never be rebound to another node. This means that some kind of controller is necessary in order for a pod to survive node failure, even if just a single pod at a time is desired.

**Comment by mboersma** [comment]
I don't think there's any debate about RC: generally, you don't even write a pod manifest, just a ReplicationController. My opinion on the service manifest is that it's useful to have one provided by default. Otherwise I'll just have to write one to actually use what the container provides, and while that service definition may often need to be custom, having a default manifest is helpful.

File: docker-registry/manifests/docker-registry-svc.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: docker-registry
+  labels:
+    app: registry
+    heritage: helm
+spec:
+  ports:
+  - port: 5000
```
Comment (slack) [inline docker-registry/manifests/docker-registry-svc.yaml]: Do we want to serve this on port 80? Seems like that might be expected in the context of a cluster service.

Comment (sgoings) [inline docker-registry/manifests/docker-registry-svc.yaml]: The Docker client communicates with Registry on 5000, does it not?

**Comment by technosophos** [comment]
@sgoings wins the Golden Documentation Digger Award :trophy: for finding the right answer to the RC question.

File: docker-registry/manifests/docker-registry-svc.yaml (hunk 2 above)
Comment (slack) [inline docker-registry/manifests/docker-registry-svc.yaml]: I always assumed 5000 was just for the "dev registry" mode, but now I'm not sure!

Comment (bacongobbler) [inline docker-registry/manifests/docker-registry-svc.yaml]: Docker uses standard URLs, so by default it is port 80 when we use `docker pull docker-registry` from within the cluster. We can map port 5000 from a pod selector (in a service) to port 80, but the replication controller should stay the same.

With this setup we'd have to do `docker pull docker-registry:5000` but the former looks a lot cleaner IMO :)

Comment (bacongobbler) [inline docker-registry/manifests/docker-registry-svc.yaml]: this is how we can do that:

```
spec:
  ports:
  - port: 80
    targetPort: 5000
```

**Comment by sgoings** [comment]
Do we want to make the volume mount required for this Chart?

**Comment by bacongobbler** [comment]
I think it's fine that the file system is ephemeral for now.

**Comment by bacongobbler** [comment]
tested manually. LGTM.

```
$ helm install docker-registry
$ helm install alpine
$ k exec -it alpine sh
/ # apk add -U curl
[...]
/ # curl registry
curl: (6) Couldn't resolve host 'registry'
/ # # damn you alpine linux and your DNS resolver...
/ # curl registry.default.svc.cluster.local
404 page not found
/ # curl $REGISTRY_SERVICE_HOST
404 page not found
```

**Pull request merged**